### PR TITLE
Fix #1384 camel-mybatis native support

### LINKED ISSUE #1384: Mybatis Component native support
Related issues:

- 1. https://github.com/quarkusio/quarkus/issues/1958
- 2. https://github.com/quarkusio/quarkus/issues/6490

### PR DESCRIPTION
**Pull request by zhfeng**
- leverage [quarkus-mybatis](https://github.com/quarkiverse/quarkus-mybatis)
- should enable ```quarkus.mybatis.xmlconfig.enable=true``` in application.properties
- must use ```<dataSource type="QUARKUS">``` in mybatis xml configuration file
- have to use ```--report-unsupported-elements-at-runtime``` in the native mode currently and should be removed when https://github.com/quarkiverse/quarkus-mybatis/issues/10 get resolved

### CODE REVIEW

**Comment by zhfeng** [comment]
This needs a new release of [quarkus-mybatis](https://github.com/quarkiverse/quarkus-mybatis)

File: docs/modules/ROOT/pages/reference/extensions/mybatis.adoc [hunk 1]
```diff
@@ -34,3 +38,25 @@ Please refer to the above links for usage and configuration details.
...
+quarkus.mybatis.xmlconfig.path=SqlMapConfig.xml
+----
+TIP: `quarkus.mybatis.xmlconfig.path` must be the same with `configurationUri` param in the mybatis endpoint.
+
+If you want to run in the native mode, you have to set these properties in pom.xml currently.
+
+[source, xml]
+----
+<properties>
+    <quarkus.package.type>native</quarkus.package.type>
```
Comment (ppalaga) [inline docs/modules/ROOT/pages/reference/extensions/mybatis.adoc:58]: `<quarkus.package.type>native</quarkus.package.type>` is not MyBatis specific and it does not need to be documented here.

File: docs/modules/ROOT/pages/reference/extensions/mybatis.adoc [hunk 2]
```diff
@@ -34,3 +38,25 @@ Please refer to the above links for usage and configuration details.
...
+    <quarkus.native.additional-build-args>--report-unsupported-elements-at-runtime</quarkus.native.additional-build-args>
```
Comment (ppalaga) [inline docs/modules/ROOT/pages/reference/extensions/mybatis.adoc:59]: Could you please explain why is this required? We generally do not want any `additional-build-args`.

File: extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java [hunk 3]
```diff
@@ -34,13 +39,19 @@ FeatureBuildItem feature() {
...
-     */
-    @BuildStep(onlyIf = NativeBuild.class)
-    @Record(value = ExecutionTime.RUNTIME_INIT)
-    void warnJvmInNative(JvmOnlyRecorder recorder) {
-        JvmOnlyRecorder.warnJvmInNative(LOG, FEATURE); // warn at build time
-        recorder.warnJvmInNative(FEATURE); // warn at runtime
+    @Record(ExecutionTime.STATIC_INIT)
+    @BuildStep
+    CamelBeanBuildItem mybatisComponent(
+            List<SqlSessionFactoryBuildItem> sqlSessionFactoryBuildItem, CamelMyBatisRecorder recorder) {
```
Comment (ppalaga) [inline extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java:45]: Why `List<SqlSessionFactoryBuildItem>`? Can there be more than one `SqlSessionFactoryBuildItem`?

File: extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java [hunk 4]
```diff
@@ -34,13 +39,19 @@ FeatureBuildItem feature() {
...
+        return new CamelBeanBuildItem("mybatis", MyBatisComponent.class.getName(),
+                recorder.createMyBatisComponent(sqlSessionFactoryBuildItem.get(0).getSqlSessionFactory()));
+    }
+
+    @Record(ExecutionTime.STATIC_INIT)
+    @BuildStep
+    CamelBeanBuildItem mybatisBeanComponent(
+            List<SqlSessionFactoryBuildItem> sqlSessionFactoryBuildItem, CamelMyBatisRecorder recorder) {
+        return new CamelBeanBuildItem("mybatis-bean", MyBatisBeanComponent.class.getName(),
+                recorder.createMyBatisBeanComponent(sqlSessionFactoryBuildItem.get(0).getSqlSessionFactory()));
```
Comment (ppalaga) [inline extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java:55]: Can we be sure that sqlSessionFactoryBuildItem is not empty?

File: docs/modules/ROOT/pages/reference/extensions/mybatis.adoc (hunk 2 above)
Comment (ppalaga) [reply docs/modules/ROOT/pages/reference/extensions/mybatis.adoc:59]: OK, now I see there is https://github.com/quarkiverse/quarkus-mybatis/issues/10 . I'd prefer not merging this unless https://github.com/quarkiverse/quarkus-mybatis/issues/10 is fixed, or unless there is a clear idea how it can be fixed.

File: integration-tests/mybatis/src/main/java/org/apache/camel/quarkus/component/mybatis/it/MybatisResource.java [hunk 5]
```diff
@@ -61,4 +77,44 @@ public Response loadComponentMybatisBean() throws Exception {
         LOG.warnf("Could not load [%s] from the Camel context", COMPONENT_MYBATIS_BEAN);
         return Response.status(500, COMPONENT_MYBATIS_BEAN + " could not be loaded from the Camel context").build();
     }
+
```
Comment (ppalaga) [inline integration-tests/mybatis/src/main/java/org/apache/camel/quarkus/component/mybatis/it/MybatisResource.java:80]: Could you please remove the original auto-generated test ensdpoints `loadComponentMybatis()` and `loadComponentMybatisBean()`?

File: integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                .statusCode(200);
+    }
+
+    @Test
+    public void loadComponentMybatisBean() {
+        /* A simple autogenerated test */
+        RestAssured.get("/mybatis/load/component/mybatis-bean")
+                .then()
+                .statusCode(200);
+    }
```
Comment (ppalaga) [inline integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java:47]:
```suggestion
```

Please remove the auto-generated tests

File: integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                .statusCode(200)
+                .body(equalTo("3"));
+    }
+
+    @Test
+    public void testDelete() {
+        RestAssured.delete("/mybatis/deleteOne?id=456")
+                .then()
+                .statusCode(200)
+                .body(equalTo("1"));
```
Comment (ppalaga) [inline integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java:89]: Running `testDelete()` before `testSelectOne()` would make `testSelectOne()` fail. You should not assume any particular test method execution order. You may want to move this code to `testSelectOne()`.

**Review by ppalaga — CHANGES_REQUESTED** [review]
I vote for merging this only if `--report-unsupported-elements-at-runtime` is not required.

File: extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java (hunk 3 above)
Comment (zhfeng) [reply extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java:45]: yeah, this was introduced to support multi datasource support see https://github.com/quarkiverse/quarkus-mybatis/pull/68

File: extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java (hunk 4 above)
Comment (zhfeng) [reply extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java:55]: sure, I create https://github.com/quarkiverse/quarkus-mybatis/issues/74 for improving it when using mybatis xml configuration.

File: integration-tests/mybatis/src/main/java/org/apache/camel/quarkus/component/mybatis/it/MybatisResource.java (hunk 5 above)
Comment (zhfeng) [reply integration-tests/mybatis/src/main/java/org/apache/camel/quarkus/component/mybatis/it/MybatisResource.java:80]: yeah, I will do.

File: integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java (hunk 7 above)
Comment (zhfeng) [reply integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java:89]: The MyBatisResource uses @Transactional and marks the transaction with rollback only. So all of the changes will not persist in the database.

**Comment by zhfeng** [comment]
Thanks @ppalaga for reviewing and I agree to not merge this PR until https://github.com/quarkiverse/quarkus-mybatis/issues/10 get resolved. I just raise it for getting some feedback and make sure the mybatis xml configuration works with camel-mybatis.

File: extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java (hunk 4 above)
Comment (ppalaga) [reply extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java:55]: I did not ask for not creating sqlSessionFactory for each data source. (I have no idea whether that makes sense for Quarkivese MyBatis at all). I asked about the case when the list of sqlSessionFactoryBuildItems is empty. Is it made impossible somewhere in Quarkivese MyBatis or here? Because otherwise `sqlSessionFactoryBuildItem.get(0)` may throw an `ArrayIndexOutOfBoundsException` or similar and it would not inform the end user that (and where) he needs to define a datasource.

Comment (zhfeng) [reply extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java:55]: OK, I get it and I think it should be in Quarkivese MyBatis to make sure to create at least one sqlSessionFactory. Also do you know if it is possible to set ```quarkus.mybatis.xmlconfig.enable=true``` automatically when the ```camel-quarkus-mybatis``` extension is activated ?

File: integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java (hunk 7 above)
Comment (ppalaga) [reply integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java:89]: I have not noticed that, thanks for the explanation. However, to ensure that the persistence layer really works, I think it would be better to test with a non-rollback transaction manager and always do a followup read call to ensure that the insert/update/delete was successful.

File: extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java (hunk 4 above)
Comment (ppalaga) [reply extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java:55]: I think the cleanest way would be to have a dedicated XmlConfigEnabledBuildItem for that in Quarkivese MyBatis. Other options I can think of are hacks.

Comment (zhfeng) [reply extensions/mybatis/deployment/src/main/java/org/apache/camel/quarkus/component/mybatis/deployment/MybatisProcessor.java:55]: Thanks @ppalaga and I will try this idea in https://github.com/quarkiverse/quarkus-mybatis/issues/84

File: integration-tests/mybatis/src/main/java/org/apache/camel/quarkus/component/mybatis/it/MybatisResource.java (hunk 5 above)
Comment (zhfeng) [reply integration-tests/mybatis/src/main/java/org/apache/camel/quarkus/component/mybatis/it/MybatisResource.java]: Done

File: integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java (hunk 6 above)
Comment (zhfeng) [reply integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java]: Done

File: integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java (hunk 7 above)
Comment (zhfeng) [reply integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java]: I just move all of the tests in one @Test method so them can run orderly

Comment (ppalaga) [reply integration-tests/mybatis/src/test/java/org/apache/camel/quarkus/component/mybatis/it/MyBatisTest.java]: > I just move all of the tests in one @test method so them can run orderly

Yeah, thanks, that's what we mostly do in case like this one.

**Comment by zhfeng** [comment]
@ppalaga I got some progress on https://github.com/quarkiverse/quarkus-mybatis/issues/10 and I think it could be fixed with https://github.com/mybatis/mybatis-3/pull/2284. So I'm waiting for the new mybatis release of 3.5.8 and come back to work with this PR.

**Comment by zhfeng** [comment]
quarkus-mybatis 0.0.11 bump mybatis to 3.5.8 and it still needs to bump mybatis version in camel-3.14.x refer to https://github.com/apache/camel/pull/6632

**Comment by ppalaga** [comment]
@zhfeng could you please rebase on top recent main to get rid of the ipfs failures?

**Comment by zhfeng** [comment]
@ppalaga can we go for merging it since CI looks good ?

**Comment by ppalaga** [comment]
> @ppalaga can we go for merging it since CI looks good ?

Yes, thank you very much for the contribution!
